### ISSUE TITLE
Question about the building scripts, for packing riscv-none-embed for Arch Linux

### ISSUE THREAD

**Issue by sequencer** [gnu-mcu-eclipse/riscv-none-gcc-build#19]
`GCC_ARCH` and `GCC_ABI` are defined here https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/109e7be4bb98ddce63bc246d6eb73bee113ef454/scripts/defs-source.sh#L37
and used in here https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/f60b9b5c52c84a9d60dd34c657ee03cb9a4cfb4f/scripts/container-apps-functions-source.sh#L389
My question is that is this gcc support all RISC-V ABI and ARCH, and these is the default settings?
And is `MULTILIB_FLAGS` void in current build script?
And `newlib` is `nano` or not in default?

**Comment by ilg-ul** [comment]
The default settings are used if you do not define them on the compiler command line.

The MULTILIB list is not void, it is quite large, and includes most reasonable combinations.

The default is the large newlib, if you want the nano version you must explictly ask for it.

**Comment by sequencer** [comment]
Got it, may I ask where can I find the `MULTILIB` generation script?

**Comment by ilg-ul** [comment]
https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/109e7be4bb98ddce63bc246d6eb73bee113ef454/scripts/container-apps-functions-source.sh#L313

**Comment by sequencer** [comment]
Got it! Thank you!

**Comment by ilg-ul** [comment]
You're welcome. Please close the issue.

**Comment by sequencer** [comment]
I'm packaging `riscv-none-embed` for Arch Linux. If I encounter more questions, can I ask here, or I should open another issue?

**Comment by ilg-ul** [comment]
If you think the questions are standalone  and might be of interrest for others, open separate issues; if not, use this ticket.

For reference, could you post here a link to the Arch package build file?

Do you also pack binutils and gdb? If so, please post links to them too.

**Comment by sequencer** [comment]
https://github.com/sequencer/pkgbuilds-riscv-none-embed
Here is the repo, I'll finish the PKGBUILDS, and then submit to a TU and push to community.

**Comment by ilg-ul** [comment]
I guess you copied the list of multilibs from a previous version, now it includes more combinations:

https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/1e710ac5956c7187a6569f9d8036e4474823259a/scripts/container-build.sh#L534

also, you enabled only c, probably you want c++ too.

**Comment by sequencer** [comment]
The reason I only enable gcc is it is only stage1.
Form the experience of Linux toolchain, I think the bootstrap sequence is:
`binutils`->`gcc-stage1`->`newlib-stage1`->`gcc-stage2`->`newlib-stage2`
I don't know if `newlib` need stage2 or not.
So the `gcc-stage1` might only need to enable C to compile the `newlib`

**Comment by ilg-ul** [comment]
ah, ok, it is still work in progress. please let me know when ready.

**Comment by sequencer** [comment]
OK, thank you for your help.

**Comment by sequencer** [comment]
You said 
https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/f60b9b5c52c84a9d60dd34c657ee03cb9a4cfb4f/scripts/container-apps-functions-source.sh#L496
Is special for ARM
But enable in
https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/f60b9b5c52c84a9d60dd34c657ee03cb9a4cfb4f/scripts/container-apps-functions-source.sh#L536
is that right? or just a typo?

**Comment by ilg-ul** [comment]
I'm not sure I can folow you. the two enable were added to the non-nano, and the disable was added to the nano.

you might want to check the original ARM script (https://github.com/gnu-mcu-eclipse/arm-gcc-original-scripts/blob/386aa44b9ce3dd321f4a572577e137261688b661/build-toolchain.sh#L348) and the original SiFive script (https://github.com/sifive/freedom-tools/blob/f096c82d31cea745dde382e185a26bcd17a6afe8/Makefile#L334)

I tried to follow them, but I'm not sure I succeeded, so if you notice differences, please let me know.

**Comment by sequencer** [comment]
I found SiFive script from here
https://github.com/sifive/freedom-tools/blob/f096c82d31cea745dde382e185a26bcd17a6afe8/Makefile#L307
is only enable
```
--enable-newlib-io-long-double \
--enable-newlib-io-long-long \
--enable-newlib-io-c99-formats \
```
But `riscv-none-embed` is with
```
--enable-newlib-io-long-double \
--enable-newlib-register-fini \
--enable-newlib-retargetable-locking \
--disable-newlib-supplied-syscalls \
--disable-nls \
--enable-newlib-io-long-long \
--enable-newlib-io-c99-formats 
```
I'm not sure if that is on purpose.
And I tried to use
`-O2` from standard `newlib` build, and `-Os` for `nano-newlib` build.
This setting might more fit for the embedded use case.

**Comment by ilg-ul** [comment]
I guess the 3 extra options were copied from the ARM script. originally SiFive did not have a newlib-nano, and I added it based on the ARM script. I don't know what to say, I would trust ARM more than SiFive. if you check what these newlib options do, and discover any resons to remove them, I'm ready to reconsider.

> `-O2` from standard newlib build, and `-Os` for nano-newlib build.

isn't the script doing exactly what you expect?

https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/f60b9b5c52c84a9d60dd34c657ee03cb9a4cfb4f/scripts/container-apps-functions-source.sh#L450

**Comment by sequencer** [comment]
And I have a question of final stage of gcc from here:
https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/f60b9b5c52c84a9d60dd34c657ee03cb9a4cfb4f/scripts/container-apps-functions-source.sh#L821
It used to think it should be pointed to `newlib` header files, the `yes` here didn't specify where is the `newlib` header path, I found the gcc with `glibc` [build script](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/riscv64-linux-gnu-glibc#n51)
has point it `/usr/$target/include`.
And It also leaks of some features comparing with [arm gcc](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/arm-none-eabi-gcc). I'm not sure if it is OK in the embedded use case.

**Comment by sequencer** [comment]
By the way, is there any plan to upstream the `riscv-none-embed` to gnu?
If I can use the upstream version of these toolchain, it will be much better for a Linux distribution.

**Comment by ilg-ul** [comment]
> --with-headers=yes

one fundamental design requirement for the xPacks is to allow multiple versions of the same package to be installed in parallel; thus, the binary distributions should not be bound to a certain absolute location; the package should be self contained in a folder and should continue to run the same when the folder is moved around.

the ARM distribution does the same:

https://github.com/gnu-mcu-eclipse/arm-gcc-original-scripts/blob/386aa44b9ce3dd321f4a572577e137261688b661/build-toolchain.sh#L392

> upstream riscv-none-embed to gnu?

yes, that would be nice, but I track the SiFive fork, which includes some features like additional attributes, and as long as they do not upstream their work, I cannot do it either.

however, if you would like to help, you could try to upstream the '-embed' patch to the 'configure.sub' project, it is only one line, and would simplify future maintenance, since now I have to patch each new version. if you agree, please write me on my personal address, to give you the details.

**Comment by sequencer** [comment]
OK! I I’ll package these packages firstly, and give a phantom package with symlink for RISCV environment.

**Comment by ilg-ul** [comment]
> I’ll package these packages firstly, and give a phantom package with symlink for RISCV environment.

I'm not sure I understand exactly what this means, but I'll do my best when ready.
 
for the next release I'm considering adding support for semihosting, using the same `--specs=rdimon.specs` option as the `arm-none-eabi` toolchain. the code will be in a separate `librdimon.a`, the SiFive `libgloss.a` should not be affected. what do you think about this?

**Comment by sequencer** [comment]
> I'm not sure I understand exactly what this means, but I'll do my best when ready.

Sorry, I forget to explain, I'm a CPU architecture student and working on rocket-chip, beside providing a clean RISC-V software toolchain, this software will be a part of [rocket-tools](https://github.com/freechipsproject/rocket-tools/), which is used to deliver a RISC-V software toolchain for CPU design.
Because of the badly maintained SiFive branch, I wanna create a phantom package, which only contains symlinks from `riscv64-unknown-elf-` to `riscv-none-embed-`.

It seems that you have already done the semihosting in the [riscv-openocd](https://github.com/riscv/riscv-openocd/pull/266), but I'm not familiar with the semihosting with openocd, so sorry about this.
But if semihosting is ready for this toolchain, I think it can be a great process for replacing origin branch.

**Comment by sequencer** [comment]
By the way from [here](https://github.com/riscv/riscv-wiki/wiki/RISC-V-Software-Status), I found that `binutils` `gcc` `newlib` has finished the upstream work, should we change to the upstreamed version, and provide a patch file to provide `riscv-none-embed`?

**Comment by ilg-ul** [comment]
> I wanna create a phantom package, which only contains symlinks from riscv64-unknown-elf- to riscv-none-embed-.

you can do this, but existing make files that use `riscv64-unknown-elf-` will probably fail, because, as far as I can tell, they run on various simulators, which require libgloss.

**Comment by sequencer** [comment]
I understand, I only need to run the project in rocket-tools, which is easy to patch.

**Comment by ilg-ul** [comment]
> I only need to run the project in rocket-tools, which is easy to patch.

if so, I suggest you patch the prefix to `riscv-none-embed-` together with adding `-lgloss`, and **do not** provide links to `riscv64-unknown-elf-`, since you add to the confusion.

maybe I was not explicit enough, but the two toolchains are not identical, and cannot be used interchangeably; regardless of what SiFive claims, as long as it ends system calls with `ecall` instructions, `riscv64-unknown-elf-` is **not** intended to run on bare metal, but on their simulators, while `riscv-none-embed-` was created exactly to run on bare metal, same as `arm-none-eabi-`.

> should we change to the upstreamed version, and provide a patch file to provide riscv-none-embed?

the SiFive gcc is not upstreamed, and I do not know if/when it will be.

otherwise, yes, providing patches to binutils/newlib to directly build `riscv-none-embed` from upstream would be nice, but you need to convince the maintainers of `config.sub` to accept the `-embed` part of the name. if you want to do it, I can help you.

**Comment by sequencer** [comment]
In current rocket-tools state, most of it is CPU spec test without privilege test, so `riscv-none-embed` is suitable for these projects. And I'll not create it in `/usr/bin`. This wrapper will only valid after activate the `$RISCV` environment. So this two toolchain can exist in the same time.
Can I construe that `riscv64-unknown-elf-` with `libgloss` will use system calls `ecall` which need OpenSBI or bbl to handle the trap? while `riscv-none-embed-` doesn't.

**Comment by ilg-ul** [comment]
> two toolchain can exist in the same time.

if, for any reason, you also need `riscv64-unknown-elf-`, you should build it separately from sources, not link to `riscv-none-embed-`.

> that `riscv64-unknown-elf-` with libgloss will use system calls ecall which need OpenSBI or bbl to handle the trap? while riscv-none-embed- doesn't.

that's correct. running applications created with `riscv64-unknown-elf-` on bare metal hangs the CPU, since the ecall handler is not present.

**Comment by sequencer** [comment]
> that's correct. running applications created with riscv64-unknown-elf- on bare metal hangs the CPU, since the ecall handler is not present.

Got it.
I wanna explain the how do I use `riscv-none-embed-`:
In the [rocketchip](https://github.com/freechipsproject/rocket-chip/) need the environment of [rocket-tools](https://github.com/freechipsproject/rocket-tools/blob/master/README.md) `$RISCV`, which contains GNU toolchain, [spike](https://github.com/riscv/riscv-isa-sim/)(simulator), [riscv-tests](https://github.com/riscv/riscv-tests/), [riscv-opcodes](https://github.com/riscv/riscv-opcodes/), all of these will not generate `ecall` to hang CPU.
`riscv-none-embed-` suit these project better than the `riscv64-unknown-elf-`.
As for Linux use case, `riscv64-linux-gnu-` is packaged in Arch Linux and Debian, which use the upstream `gcc` and `glibc`, it's also better than `riscv64-unknown-elf-`.
So the last is the embedded program which need supervisor mode, they can be the payload of OpenSBI, compiled by `riscv-none-embed-`, and user can choose which `libc` should use, and `musl` or `newlib` can link to the progrom.

So I think in the common use case, `riscv64-unknown-elf` is useless.

**Comment by sequencer** [comment]
I have build `binutils`, `gcc`, and `newlib`, and try to submit it to TU, it seems that semihosting didn't support the current version, so I may build `gdb` later after semihosting finish.

**Comment by ilg-ul** [comment]
Can you confirm that you alredy tried the `riscv-none-embed-` binaries I provide for Linux and you were able to run all your build scripts, tests, etc?

**Comment by sequencer** [comment]
I have run my build scripts, but I didn't use the binaries you provided(Arch Linux doesn't have this package), I may test it after rocket-tools finished.

**Comment by ilg-ul** [comment]
support for semihosting should be added in newlib, gdb has nothing to do with it, you can build it too.

and adding semihosting to newlib might take a while, it is not a high priority for me, it'll probably be available in the next release; the one that I'll probably release tomorrow will not include it, is is only a maintenance release to fix the macOS GDB problems.

**Comment by sequencer** [comment]
OK I got it, I'm going to package `gdb`.

**Comment by sequencer** [comment]
I have a question that:
[It](https://github.com/gnu-mcu-eclipse/riscv-none-gcc-build/blob/master/README.md#checkout-remote-branches) said gdb is using the [upstream version](https://sourceware.org/git/?p=binutils-gdb), I use `ftp://ftp.gnu.org/gnu/gdb`, but I failed to compile because of no `embed` supported for gdb.
May I ask which upstream should I choose?

**Comment by ilg-ul** [comment]
> I didn't use the binaries you provided(Arch Linux doesn't have this package)

???

did you read the page on how to install my binaries?

**Comment by ilg-ul** [comment]
> failed to compile because of no embed supported for gdb.

sure, it must be added to the centralised  config.sub project and later used by all projects.

> gdb ... which upstream should I choose?

the one referred by my scripts, similar to gcc.

**Comment by ilg-ul** [comment]
as for using an Arch toolchain package for building the rocket-tools project, you might have some surprises when using a different toolchain version than the one it was already tested.

**Comment by ilg-ul** [comment]
also for using the Arch rolling distribution, how do you plan to handle the case when one project requires riscv-none-embed-gcc v7.4.0 and another one requires v8.2.1, exactly those versions, and the current version is, lets's say, 8.3?

**Comment by sequencer** [comment]
> also for using the Arch rolling distribution, how do you plan to handle the case when one project requires riscv-none-embed-gcc v7.4.0 and another one requires v8.2.1, exactly those versions, and the current version is, lets's say, 8.3?

In currently state of Arch Linux, we will follow the upstream stable version tightly, and won't use the out-of-date toolchain.
If some software has to be built by gcc 7.4, I think this software is **out-of-date**, which should be port to the latest version of gcc.
Most of toolchain maintainers' troubles are derived from the badly maintained upstream(like SiFive), and strange need from downstream(like out-of-date software).
So push toolchain into stable upstream is the most important thing(which keep the maintaining easy and safe). As for out-of-date software from downstream, Arch provide these way to resolve it. In distribution, we provide a window phase to maintain 2 version of toolchain(update version will locate in the `testing`), which let developers to port software to new toolchain version, till all package is tested, new version will replace old one. If it has too much trouble to port, like python2 to python3, we will provide both package and deprecate python2 gradually. If some user are obsess to the out-of-date version, he can modify the PKGBUILD(or write it by himself) and publish it AUR, use AUR helper to automatically package and install. As the example of `arm-none-eabi`, AUR has such gcc's:
```
aur/arm-none-eabi-gcc47-linaro 4.7_2014.06-1 [1 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc47-linaro-alternative 4.7_2014.06-1 [0 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc48-linaro 4.8_2014.08-1 [1 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc49-linaro 4.9_2016.02-1 [3 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc49-linaro-alternative 4.9_2016.02-1 [0 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc53-linaro 5.3_2016.02-1 [1 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc53-linaro-alternative 5.3_2018.11-1 [0 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc55 5.5.0-1 [0 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target
aur/arm-none-eabi-gcc60-linaro 6.4_2018.04-1 [1 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc64-linaro-alternative 6.4_2017.06-1 [0 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc73-linaro 7.3_2018.05-1 [0 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
aur/arm-none-eabi-gcc74-linaro 7.4_2019.02-1 [0 / 0.000000]
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target.
```
But in the official repo:
```
community-testing/arm-none-eabi-gdb 8.3-1
    The GNU Debugger for the ARM EABI (bare-metal) target
community/arm-none-eabi-binutils 2.32-1
    A set of programs to assemble and manipulate binary and object files for the ARM EABI (bare-metal) target
community/arm-none-eabi-gcc 9.1.0-1
    The GNU Compiler Collection - cross compiler for ARM EABI (bare-metal) target
community/arm-none-eabi-gdb 8.2.1-2
    The GNU Debugger for the ARM EABI (bare-metal) target
community/arm-none-eabi-newlib 3.1.0-1
    A C standard library implementation intended for use on embedded systems (ARM bare metal)
```
We can find gcc 9.1 is in the stable repo, and only gdb 8.3 is under testing(the window phase).

We keep PKGBUILD easy to modify(one file to build), make each user easy to package his own package, and we keep our software as newest as possible(to provide new features), and avoid using upstream dev branch(fedora always does it) to introduce bugs.
As the result, the main work from packager is to fire the issue to upstream to report bugs, and fire the issue to downstream to keep some lazy ones follow the upstream.

**Comment by sequencer** [comment]
After solving the gdb problem, I'll upstate this package to AUR. And we can together focus on the `embed` upstreaming.
After finish upstream, I can ask TU to move `riscv-none-embed` from AUR to community and follow upstream to provide a clean embed toolchain like `arm-none-eabi-`.

**Comment by ilg-ul** [comment]
> In currently state of Arch Linux, we will follow the upstream stable version tightly, and won't use the out-of-date toolchain.
> If some software has to be built by gcc 7.4, I think this software is out-of-date, which should be port to the latest version of gcc.
> Most of toolchain maintainers' troubles are derived from the badly maintained upstream(like SiFive), and strange need from downstream(like out-of-date software).
> So push toolchain into stable upstream is the most important thing(which keep the maintaining easy and safe). As for out-of-date software from downstream, 

ok, if you are happy with this philosophy, fine. 

it might work for hacking, but in real life you'll notice there are tons of production projects that remain stuck their entire life to the initial toolchain version, possibly a decade old.

as for the RISC-V toolchain, the plan for `riscv-none-embed` is to follow the SiFive releases, which do not come very often, so you'll probably not be as up to date as you wish.

if you manage to upstream the `-embed` patch you'll be able to make your own intermediate releases and do not wait for my forks to be updated. but don't expect this to be easy, SiFive never liked the idea of `riscv-none-embed` and may oppose to the patch.

btw, I'm still curious if you can fully built rocket-tools with the new `riscv-none-embed` toolchain.

**Comment by sequencer** [comment]
I will talk with Palmer face to face in June, at Shenzhen, China.
Hope luck with me.
I'm considering using patch with upstream to follow the upstream, it that possible?

**Comment by ilg-ul** [comment]
> Hope luck with me.

sure, you'll need it!

> I'm considering using patch with upstream to follow the upstream, it that possible?

I don't know what this means in practical terms, but feel free to use any solutions you think apropriate.

from my point of view, there are two patches needed:

- one to the config.sub project, to enable `-embed` as an OS name, to be used temporarily until a RISC-V EABI will be defined and we'll be able to use `riscv-none-eabi`; however please note that I do not know of any active efforts to define the EABI, so this might take years.

- the second to remove the mandatory -lgloss from the specs.

if you manage to push these patches upstream, I'll probably contribute the semihosting code to newlib, as the rdimon library, similarly to arm-none-eabi.

**Comment by sequencer** [comment]
Currently, is there any suitable abi for `eabi`? Or any candidates `eabi` spec till now? 
Or just use psabi as `eabi` and remove the libgloss?

**Comment by ilg-ul** [comment]
As a starting point I suggested https://github.com/emb-riscv/specs-markdown/blob/develop/eabi.md, but was silently ignored.

**Comment by sequencer** [comment]
I just found that. But I wonder if it is too hard to upstream a new abi to gcc and clang?

After checking the prefix name rule, I think `none` and `embed` has the same meaning, and didn't specify which `abi` are we using. So I wonder if we wanna reduce the maintaining trouble, and upstream as less as we can:
If it is possible to change prefix of `riscv-none-embed-` to `riscv-none-psabi-`, and get rid of the libgloss?(just assume, I have no experience in embed programming, and gnu toolchain maintaining), or just use `riscv-none-eabi` to avoid porting problem after the new `eabi` merged by upstream?

These are my personal suggestion, I'm not sure it's OK or not.

**Comment by sequencer** [comment]
Next time, after I analyses difference ABIs, I will propose this topic inside SiFive China, and try to get more attentions.

**Comment by ilg-ul** [comment]
> I wonder if it is too hard to upstream a new abi to gcc and clang?

yes. implementing a new ABI is always a serious project, even if it can be inspired from an existing related ABI.

but this is not enough. 

probably even more difficult is to make the decision how to use the registers efficiently, given the hardware specifics, like compressed mode. this means a very, very good understanding of the instruction set, followed by implementing several versions of the ABI in the compiler and running countless builds to compare the resulting code size and speed. this is by all means not an easy task.

> `riscv-none-psabi-`

sorry, from my point of view this is a nonsense, I cannot support it.

> `riscv-none-eabi-`

cannot support this either, not until we'll really have an EABI.

adding `-embed` is actually very simple, a single line must be edited, a line like:

https://github.com/gcc-mirror/gcc/blob/90b483b4c348b6915e19b493851dcc5d1581c87e/config.sub#L1354

`embed` can be considered the name of an embedded operating system, like `rtems`; at least this is how I would phrase the request to edit this line.

**Comment by sequencer** [comment]
I used to think this area is for `abi`, and after reading the `config.sub`, I agree with you, basic system types is not a `psabi`, and `eabi` is fake until the true `eabi` comes out.
What do you think the difference between `embed` and `elf`?
`embed` is truly working like a kind of OS(all task handled by program ifself), and `elf` is also executable and linkable file which contains the same meaning. So if `elf` is already upstreamed, why should we upstream `embed` with the same meaning?

**Comment by ilg-ul** [comment]
> What do you think the difference between embed and elf?

`-elf` was intended to produce ELF images that run inside a simulator, or on top of a monitor, via libgloss. SiFive calls this _bare metal_, but this is not accurate.

`-embed`, as a temporary replacement for `-eabi`, produces binaries that are intended to run on bare metal, generally from a flash memory.

**Comment by sequencer** [comment]
In my heart. I totally agree with the `embed`, because `elf` is not an os.
But `embed` is not upstreamed and packager always need a patch to maintain.
If you are willing to help me with the upstream procedure, I’ll be grateful and willing to do so.

**Comment by sequencer** [comment]
I got it. 
My next todo is to make a minimal patch to riscv-gcc to make it more easy to package.
And we can together work on the upstream procedures.

**Comment by ilg-ul** [comment]
> a minimal patch to riscv-gcc

I did not check, but the SiFive specific changes (like to add specific interrupt attributes) may not be minimal at all.

and of course I'll help you with the upstream procedures.

**Comment by ilg-ul** [comment]
I tried to send you a message on the private email address you used when registering to GitHub. did you receive it?

**Comment by sequencer** [comment]
Sorry for my late reply, my house was flooded these days, and I was busy dealing this :(
But I didn't receive the message in my mailbox, and I have send a new email to you, making you can receive this.

**Comment by sequencer** [comment]
With TU(@felixonmars) by my side, checking the PKGBUILDs:
  1. we can maintain a patch of upstream version, using this to patch upstream and `riscv-gcc`(which can use the recently commit by SiFive changes).
  2. we can firstly send this package to AUR, after upstreaming, it can be moved to community repository.

**Comment by sequencer** [comment]
I have checked the [riscv-gcc](https://github.com/riscv/riscv-gcc/), there is no recently change on that repository, and seem this repository is stale now.

> I did not check, but the SiFive specific changes (like to add specific interrupt attributes) may not be minimal at all.

I didn't find the related commits, is that they use gcc 8.3.x or 8.2.x version, adding these features without upstream them?

**Comment by sequencer** [comment]
I found a **new** patch  from jimw [here](https://patchwork.ozlabs.org/patch/940251/) is that a good start for this?

**Comment by ilg-ul** [comment]
> I have checked the riscv-gcc

if you check the readme files in my build project, or the comments in the build script, you'll find the URL of the SiFive GCC repository, which is up to date.

> I found a new patch from jimw here is that a good start for this?

originally the config.sub script included support only for `riscv32-*-*` or `riscv64-*-*`, and I had to apply a patch there to allow `riscv-*-*`. this patch from jim done exactly this, so we have to patch only the `-embed` part.

**Comment by sequencer** [comment]
I use you patches to patch `config.sub` with `embed` in OS and remove `libgloss` from the GNU upstream.
I will push it to Arch Linux CN [repo](https://github.com/archlinuxcn/repo/pull/1192), and have some wrapper in the feature. Hope it can work as excepted.

BTW, I noticed that eabi spec has been added to official repo of RISC-V by Krste. Hope things can progress smoothly.

**Comment by ilg-ul** [comment]
> eabi spec repo

yes, since last night. when the new eabi will be fully functional, I'll rename the toolchain to `riscv-none-eabi`.

**Comment by sequencer** [comment]
I found most of SiFive/gcc commits in 8.2.0 has been upstreamed.
In the newest `9.1.0`, only clic related commit not ready to upstream, and new `short forward branch optimization` has been uploaded but not merged.
So If we can maintain a patch sets to choose which patch should be apply to upstream can save a lot of dirty works.

**Comment by sequencer** [comment]
> -elf was intended to produce ELF images that run inside a simulator, or on top of a monitor, via libgloss. SiFive calls this bare metal, but this is not accurate.

I'm not sure your opinion about `-elf` is correct or not, I just checked some official information from [FSF](https://sourceware.org/autobook/autobook/autobook_17.html),
> Configuration names may be used to describe all sorts of systems, including embedded systems which do not run any operating system. In this case, the field is normally used to indicate the object file format, such as ‘elf’ or ‘coff’. 

and I think `-elf` is correct here, and `-embed` can also mean `-elf` too.
But `libgloss` is certainly a mistake from SiFive, which should be fixed.
Also in my distribution, unlike SiFive's `unknown`, I decide to add `sifive` to manufacturer domain when applying CLIC and U7 support patch.
So my package might separate from `riscv-none-embed` with a name in `riscv-sifive-elf`, so don't worry about the problem that I might introduce different behavior between mine and yours.

**Comment by ilg-ul** [comment]
> I'm not sure your opinion about -elf is correct or no

I rephrase: '-elf was intended **by SiFive** to produce ELF images that run inside a simulator, or on top of a monitor, via libgloss'

So it is my opinion on what SiFive opinion is ;-)

personally I would use `-eabi` if available, and I would not had a problem with `-elf` if libgloss was not automatically linked. `-embed` was a compromise. I will stop using it as soon as EABI will be available.

> riscv-sifive-elf

it is your choice, but too many toolchain names simply increase entropy.

from my point of view, if you use `riscv-none-embed`, I would expect the functionality to be the same.
